pinot-core/src/main/java/com/linkedin/pinot/core/periodictask/PeriodicTaskScheduler.java

```diff
@@ -0,0 +1,124 @@
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Periodic task scheduler will schedule a list of tasks based on their initial delay time and interval time.
+ */
+public class PeriodicTaskScheduler {
+  private static final Logger LOGGER = LoggerFactory.getLogger(PeriodicTaskScheduler.class);
+  private final ScheduledExecutorService _executorService;
+  private long _initialDelayInSeconds;
+  private volatile boolean _running;
+
+  private static class PeriodicTaskEntry implements Comparable<PeriodicTaskEntry> {
+    private PeriodicTask _periodicTask;
+    private long _executionTime;
+
+    PeriodicTaskEntry(PeriodicTask periodicTask) {
+      _periodicTask = periodicTask;
+      _executionTime = System.currentTimeMillis() + _periodicTask.getInitialDelayInSeconds() * 1000L;
+      if (_periodicTask.getIntervalInSeconds() > 0L) {
+        _periodicTask.init();
+      }
+    }
+
+    PeriodicTask getPeriodicTask() {
+      return _periodicTask;
+    }
+
+    long getExecutionTime() {
+      return _executionTime;
+    }
+
+    void updateExecutionTime() {
+      _executionTime = System.currentTimeMillis() + _periodicTask.getIntervalInSeconds() * 1000L;
+    }
+
+    @Override
+    public int compareTo(PeriodicTaskEntry o) {
+      if (this._executionTime == o._executionTime) {
+        return 0;
+      }
+      return this._executionTime < o._executionTime ? -1 : 1;
+    }
+  }
+
+  public PeriodicTaskScheduler(long initialDelayInSeconds) {
+    LOGGER.info("Initializing PeriodicTaskScheduler. Initial delay in seconds: {}", initialDelayInSeconds);
+    _executorService = Executors.newScheduledThreadPool(1);
+    _initialDelayInSeconds = initialDelayInSeconds;
+    _running = true;
+  }
+
+  /**
+   * Start scheduling periodic tasks.
+   */
+  public void start(List<PeriodicTask> periodicTasks) {
+    if (periodicTasks == null || periodicTasks.isEmpty()) {
+      LOGGER.warn("No periodic task assigned to scheduler!");
+      return;
+    }
+
+    LOGGER.info("Starting PeriodicTaskScheduler.");
+    PriorityBlockingQueue<PeriodicTaskEntry> queue = new PriorityBlockingQueue<>(periodicTasks.size());
+    for (PeriodicTask task : periodicTasks) {
+      queue.offer(new PeriodicTaskEntry(task));
+    }
+
+    // Set up an executor that executes tasks periodically
+    _executorService.schedule(() -> {
+      while (_running) {
+        PeriodicTaskScheduler.PeriodicTaskEntry taskEntry = null;
+        try {
+          taskEntry = queue.take();
+          long currentTime = System.currentTimeMillis();
+          if (currentTime < taskEntry.getExecutionTime()) {
+            Thread.sleep(taskEntry.getExecutionTime() - currentTime);
+          }
+          LOGGER.info("Running Task {}", taskEntry.getPeriodicTask().getTaskName());
+          taskEntry.getPeriodicTask().run();
+        } catch (InterruptedException ie) {
```

Comment:
Should we ignore the interrupt? We have a global Throwable catch block

Reply:
I think it's essential to keep it there. One the one hand, this allows the interrupt to propagate - the method should not continue on ignoring the interrupt, which is beneficial for us to test. On the other hand, the global throwable block just simply logs the exception/error without stopping it, so that it won't block the next following task.

Reply:
**Thread.currentThread().interrupt();** has no effect because the thread will terminate just after this.